nodes/viz/viewer2d.py

```diff
@@ -0,0 +1,744 @@
+        # fullList(vector_color, len(p))
+        # fullList(path_color, len(p))
+
+
+        # end early
+
+
+        if self.activate:
+
+            config = lambda: None
+            x, y, scale, multiplier = self.get_drawing_attributes()
+            margin = 10* scale
+            config.mode = self.mode
+            config.loc = (x, y - margin)
+            config.palette = palette_dict.get(self.selected_theme_mode)[:]
+            config.sys_scale = scale
+            config.scale = scale*self.draw_scale
+            config.cyclic = self.cyclic
+            config.background_color = self.background_color
+            config.plane = self.working_plane
+            config.vector_color = vector_color
+            config.edge_color = path_color
+            config.draw_verts = self.vector_toggle
+            config.draw_edges = self.edge_toggle
+            config.draw_polys = self.polygon_toggle
+            config.point_size = self.point_size
+            config.edge_width = self.edge_width
+            config.color_per_point = self.color_per_point
+            config.color_per_edge = self.color_per_edge
+            config.color_per_polygon = self.color_per_polygon
+            config.polygon_use_vertex_color = self.polygon_use_vertex_color
+            config.edges_use_vertex_color = self.edges_use_vertex_color
+            config.edges = edges
+
+
+            if self.mode == 'Number':
+                config.size = self.drawing_size
+                config.continuous = self.continuous
+                geom = generate_number_geom(config, p)
+            elif self.mode == 'Path':
+                geom = generate_graph_geom(config, p)
+            elif self.mode == 'Curve':
+                paths = []
+                for curve in curves:
+                    t_min, t_max = curve.get_u_bounds()
+                    ts = np_linspace(t_min, t_max, num=self.curve_samples, dtype=np_float64)
+                    paths.append(curve.evaluate_array(ts).tolist())
+
+                geom = generate_graph_geom(config, paths)
+
+            else:
+                if not self.inputs['Edges'].is_linked and self.edge_toggle:
+                    config.edges = polygons_to_edges(polygons_s, unique_edges=True)
+
+                geom = generate_mesh_geom(config, p, polygons_s, poly_color)
+
+
+
+            draw_data = {
+                'mode': 'custom_function',
+                'tree_name': self.id_data.name[:],
+                'loc': (x, y),
+                'custom_function': simple28_grid_xy,
+                'args': (geom, config)
+            }
+            nvBGL.callback_enable(n_id, draw_data)
+
+    def sv_free(self):
+        nvBGL.callback_disable(node_id(self))
+
+    def sv_copy(self, node):
+        # reset n_id on copy
+        self.n_id = ''
+
+    def sv_update(self):
+        # handle disconnecting sockets, also disconnect drawing to view?
+        if not ("Float" in self.inputs):
+            return
```

Comment:
because this is quite a dynamic node, with many sockets (hidden), and none of them are called "Float", you could instead do a length test.  Say if this node has a total of 7 sockets (in various states of hide/shown), then the first line in sv_update could be 
```python
if not len(self.inputs) == total_expected_socket_count:
    return

```

Reply:
because this function is called several times, i think each time a new socket is added to the ui (like sv_init). Best to return early when we know the node isn't entirely initialized.

Reply:
in these nodeview viewer nodes, the purpose of this code in the try/except is to make sure the node stops drawing when there are no longer any incoming links

Reply:
That part of the code was a vestige of the base node i started from (easing 0..1). But is it really necessary? the node stops drawing when i un-link the sockets... (de bgl call is disabled at the beginning of the process function)

Reply:
there are two things here:
- when the node is entirely disconnected, the sv_update is called one more time. that where the try/except gets a change to disable drawing.
- the initial "early" return is probably not that important anymore.. if the sv_update function itself is only doing very minimal processing (ie.. none in this case) , but...  if you don't check that all sockets are created you'll get some Exceptions during the sv_init (at least.. that was always the case in the old update system..i'm not sure that's changed)